cilium/CFP-27752-operator-manages-cilium-identities.md

```diff
@@ -0,0 +1,214 @@
+
+IPCache is updated when remote CEPs/CESs are received.
+
+## Label changes
+
+
+### Pod
+
+Pod/Deployment label changes result in pod recreation, which follows the same pod creation path previously described.
+
+
+### Namespace
+
+When namespace labels change, Operator and Agents get the update at the same time. Agents wait until Operator processes the namespace label change update and creates all the new CIDs. Agents keep retrying to match CIDs for all their local endpoints until they are received through the CID watcher.
+
+
+## Simultaneous CID management compatibility
+
+The only problem with simultaneous CID management (both Operator and Agents are managing CIDs) is when namespace labels are changed. The only bad thing that happens is that there is CID duplication. This is an existing problem already, that doesn't cause disruption.
+
+Simultaneous CID management is only expected to happen during upgrade from Agents managing CIDs to Operator managing CIDs.
+
+
+## CID deduplication
+
+Operator will not create duplicate CIDs, but it will allow for their existence. In case other sources, like Agents, are making duplicate CIDs, Operator will be compatible with tracking multiple CIDs. This is needed to make Operator compatible with Agents managing CIDs at the same time.
+
+
+## KVStore compatibility
+
+[KVStore](https://docs.cilium.io/en/stable/kvstore/) is a key-value store that is mostly used for sharing Cilium mappings between IPs, identities (IPCache) and labels, across the cluster ([what's inside Cilium’s KVStore](http://arthurchiao.art/blog/whats-inside-cilium-etcd/)).
+
+Operator managing CIDs is intended to work with both `identity-allocation-mode=kvstore` and `identity-allocation-mode=crd`. CRD being the default, the initial implementation will use the CRD allocation mode, and later expanded to other allocation modes.
+
+All features that rely on using kvstore will not work until CID controller is expanded to work with kvstore allocation mode. The work to enable kvstore compatibility is estimated to of medium size and it will be tracked in a separate GitHub issue (https://github.com/cilium/cilium/issues/34865).
+
+
+# Enablement
+
+The feature will be guarded by a flag `operator-manages-identities`, that can be set in container arguments in Operator’s and Agent’s manifests. The default values for both will be `false`, meaning that Operator doesn’t manage CIDs, while Agents do.
+
+
+## Long term plan
+
+The long term plan is to make Operator managing CIDs the default, and deprecate the part of Agent that is currently used to manage CIDs.
+
+
+# Migration
+
+Any situation when neither Operator nor Agents are managing CIDs is effectively a downtime for pod creation. The control plane side of network policies and other features relying on CIDs will not carry out any required changes during this time, causing them to stop working.
+
+
+## Upgrade
+
+Safe upgrade, without downtime, requires simultaneous CID management compatibility.
```

Comment:
I would suggest breaking upgrade down into the exact upgrade steps you are analysing and describe the expectations. Specifically:

Is this upgrading from a version with Agent managing CID to Operator managing CID? What about upgrading from a cluster where operator manages CID to a newer version where operator manages CID?

The sentence here sounds like it's suggesting that users should re-enable simultaneous CID management during every subsequent upgrade. But I don't think that's what you mean.

Reply:
Thanks for the suggestion. It's more clear now. Done.